gaf/gaf.1
.TH gaf 1 "December 3rd, 2012" "gEDA Project" 1.9.0.20121203
.SH NAME
gaf - gEDA/gaf Command-Line Utility
.SH "SYNOPSIS"
.B gaf
[\fIOPTION\fR ...] \fICOMMAND\fR [\fIARGS\fR ...]
.SH "DESCRIPTION"
.PP
.B gaf
is part of the gEDA (GPL Electronic Design Automation) toolset.  It
provides a number of small command-line utilities for working with
schematic and symbol files, and is designed to be used for batch
processing of designs created using the schematic editor
\fBgschem\fR(1).  It currently has three built-in \fICOMMAND\fRs:

.B gaf export
is used to create SVG, PDF, PNG, PS and EPS files from schematic and
symbol files, for printing or embedding in other documents.

.B gaf config
allows reading and writing settings in gEDA project, user and system
configuration stores.

.B gaf shell
provides a Scheme REPL for command-line batch processing of schematic
data.

.SH "GENERAL OPTIONS"
.TP 8
\fB--no-rcfiles\fR
Prevent `gafrc' Scheme initialisation files from being loaded.
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.TP 8
\fB-V\fR, \fB--version\fR
Print \fBgaf\fR version information.

.SH "EXPORTING IMAGE FILES"
.B gaf export
[\fIOPTION\fR ...] \fB-o\fR \fIOUTPUT\fR [\fI--\fR] \fIFILE\fR ...

.B gaf export
can export schematic and symbol files in a variety of image formats
for printing or further processing.  It currently supports single-page
PNG, SVG and EPS output, and multi-page PS and PDF output.  It accepts
a variety of options for controlling how the output is formatted.

.TP 8
\fB-o\fR, \fB--output\fR=\fIFILE\fR
Output generated image data to \fIFILE\fR.
.TP 8
\fB-f\fR, \fB--format\fR=(\fBpng\fR | \fBpdf\fR | \fBsvg\fR | \fBps\fR | \fBeps\fR)
Specify an output format.  Usually, this option is not required,
because \fBgaf export\fR will infer the correct format from the file
extension of the output file.
.TP 8
\fB-p\fR, \fB--paper\fR=\fINAME\fR
Size the output for a particular paper size.  The \fINAME\fR should be
a PWG 5101.1-2002 paper name.  For example, valid values include
`iso_a4', `na_letter', or `na_d'.
.TP 8
\fB-l\fR, \fB--layout\fR=(\fBauto\fR | \fBlandscape\fR | \fBportrait\fR)
When using a paper size, set the orientation of the output.  If `auto'
layout is used, the orientation that best fits the drawing will be
used.
.TP 8
\fB-s\fR, \fB--size\fR=(\fBauto\fR | \fIWIDTH\fR;\fIHEIGHT\fR)
Size the output with specific dimensions.  If the size is `auto',
select the size that best fits the drawing.
.TP 8
\fB-m\fR, \fB--margins\fR=(\fBauto\fR | \fITOP\fR[;\fILEFT\fR[;\fIBOTTOM\fR[;\fIRIGHT\fR]]])
Set the widths of the margins to be used.  If `auto' margins are
specified, a sensible default value will be chosen.  Up to four margin
widths can be provided.  If one is provided, it will be used on all
four sides.  If two are provided, the first will be used for the
top/bottom and the second for the left/right.  If three are provided,
the first will be used for the top, the second for left/right, and the
third for the bottom.
.TP 8
\fB-a\fR, \fB--align\fR=(\fBauto\fR | \fIHALIGN\fR;\fIVALIGN\fR)
Set how the drawing is aligned within the page.  \fIHALIGN\fR controls
the horizontal alignment, and \fIVALIGN\fR the vertical.  Each
alignment value should be in the range 0.0 to 1.0.  The `auto'
alignment is equivalent to a value of `0.5;0.5', i.e. centered.
.TP 8
\fB-d\fR, \fB--dpi\fR=\fIDPI\fR
Set the number of pixels per inch used when generating PNG output.
.TP 8
\fB-c\fR, \fB--color\fR
Enable colour output.
.TP 8
\fB--no-color\fR
Disable color output.
.TP 8
\fB-F\fR, \fB--font\fR=\fIFONT-FAMILY\fR
Set the font to be used for drawing text.
.TP 8
\fB--\fR
Treat all remaining arguments as schematic or symbol filenames.  Use
this if you have a schematic or symbol filename which begins with `-'.

.PP
The \fB--size\fR or \fB--margins\fR options described above accept
values using units of `mm', `cm', `in', `pc', `px', or `pt'.  If you
do not provide a unit, points are assumed.  N.b. that `px' are
evaluated relative to the current \fB--dpi\fR setting.

.SH "ACCESSING CONFIGURATION"
.B gaf config
[\fIOPTION\fR] [\fIGROUP\fR \fIKEY\fR [\fIVALUE\fR]]

.B gaf config
is a basic utility for inspecting and modifying gEDA/gaf configuration
stores.

.PP
If a \fIGROUP\fR and \fRKEY\fR are specified, retrieves the value of
that configuration parameter.  If a \fIVALUE\fR is specified, sets
the value of that parameter.  The \fB-p\fR, \fB-u\fR and \fB-s\fR
options can be used to select the configuration store affected (by
default, the project configuration store for the current working
directory).  If no \fIGROUP\fR and \fIKEY\fR are provided, outputs
the filename of the selected configuration store.

.PP
.TP 8
\fB-p\fR, \fB--project\fR[=\fIPATH\fR]
Select the project configuration store of the project located in
\fIPATH\fR.  If no \fIPATH\fR is provided, selects the project in the
current working directory.
.TP 8
\fB-u\fR, \fB--user\fR
Select the user configuration store.
.TP 8
\fB-s\fR, \fB--system\fR
Select the system configuration store.  Depending on user permissions,
the system configuration store may be read-only.

.SH "SCHEME PROCESSING"
.B gaf shell
[\fIOPTION ...\fR]

.B gaf shell
provides a Scheme Read-Eval-Print Loop (REPL) for automating
processing of schematic and symbol files.  It is designed to be used
with the gEDA Scheme API.

.TP 8
\fB-s\fR \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR, and exit.
.TP 8
\fB-c\fR \fIEXPR\fR
Evaluate the Scheme expression \fIEXPR\fR, and exit.
.TP 8
\fB--\fR
Stop scanning arguments; run interactively.
.TP 8
\fB-L\fR \fIDIRECTORY\fR
Prepend \fIDIRECTORY\fR to the list of directories to be searched for
Scheme files.
.TP 8
\fB-l\fR \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR.

.PP
The \fB-s\fR, \fB-c\fR and \fB--\fR switches stop argument processing
and pass all the remaining arguments as the value of `(command-line)'.

.SH ENVIRONMENT
.TP 8
.B GEDADATA
specifies the search directory for Scheme and rc files.  The default
is `${prefix}/share/gEDA'.
.TP 8
.B GEDADATARC
specifies the search directory for rc files.  The default is `$GEDADATA'.

.SH AUTHORS
See the `AUTHORS' file included with this program.

.SH COPYRIGHT
.nf
Copyright \(co 2012 gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBgschem\fR(1).

For more information on the Scheme API, see the \fBgeda-scheme\fR
Texinfo manual.  If the \fBinfo\fR program is properly installed at
your site, the command

.IP
.B
info geda-scheme

.PP
should give you access to the complete manual.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2012 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.1
=================================

* `gaf export' now accepts a `--scale' option.

* `gnetlist' now outputs internationalised messages.

Notable changes in gEDA/gaf 1.9.0
=================================

* A new schematic renderer library, `libgedacairo' has been added, and
  all tools have been updated to use it.  Related changes:

  - The following configuration functions have been removed:

         output-capstyle         paper-sizes
         output-color            postscript-prolog
         output-orientation      print-command
         output-type             setpagedevice-orientation
         paper-size              setpagedevice-pagesize

  - The following `gschem' configuration items have been added:

         print-paper             print-color
         print-orientation

  - `gschem' now generates PDF files rather than PostScript or EPS
    files.

  - `gschem' now uses a standard GTK print dialog (with printer
    selection, print preview, etc.)

* A new `gaf' command-line utility has been added.  It provides a
  number of subcommands:

  - `gaf export' can be used to convert schematic files to PNG, SVG,
    PDF, PS and EPS, with detailed control of paper size and margins.
    It supports multi-page PDF and PS output.

  - `gaf config' can be used to manipulate the project, user, and
    system configuration stores (see below).

  - `gaf shell' provides a Scheme REPL for batch manipulation of
    schematic and symbol files.

* New project, user and system configuration stores have been added,
  using a format similar to .desktop files.  They can be manipulated
  using the `gaf config' tool.  Configuration parameters for most
  tools will be converted to use this mechanism in future releases.

* The changes to highlighting of dangling net ends in gEDA/gaf 1.8.0
  have been reverted.  All dangling ends are once again flagged with a
  brightly-coloured square.

* When editing a component with the `gschem' multi-attribute editor,
  the dialog will now show the symbol name in its title.

* When using "Down Schematic" command in `gschem', an error message
  will now be displayed if the subcircuit schematic cannot be loaded.

* Printing now uses a light background color map by default.

* Objects can now have the line end "cap style" adjusted in `gschem'.

* `gnetlist' now accepts the `-i' and `-g' command-line arguments
  together.  If you specify both `-i' and `-g', `gnetlist' will enter
  the Scheme REPL after loading the requested backend but before
  executing it.

* `gattrib' now has translations for the user interface and various
  messages.

Notable changes in gEDA/gaf 1.8.0
=================================

* `gschem' and `gnetlist' have updated and rewritten user guides.

* All tools now give more informative error messages if they are
  unable to load schematic or symbol files due to syntax errors.

* All tools now correctly handle schematic and symbol files with `\r',
  `\n', or `\r\n` line endings, or a mixture.

* Unconnected pin ends and net ends are now all drawn with the same
  box marker in `gschem' and printed output.

* Net segments with at least two connections (including indirect
  connections via `netname=' attributes) are now drawn with arrowheads
  rather than box markers on dangling ends in `gschem' and printed
  output.

Notable changes in gEDA/gaf 1.7.2
=================================

* `geda-gaf' now requires GTK+ 2.16.0 or later.

* A greatly expanded Scheme API has been added to gEDA/gaf for use by
  extension authors.  See the `geda-scheme' Info manual for more
  details.  Existing extensions may need to be modified to load the
  `(geda deprecated)' or `(gschem deprecated)' modules.

* The `component-library-search' function for `gafrc' files now skips
  directories without symbols, and searches for libraries recursively.

* `gschem' keybinding improvements:

  - Keybindings are now unaffected by Caps Lock.

  - Keystrokes are displayed using the same characters as on keycaps
    (e.g. `:' instead of `colon').

  - Keystrokes may use extended modifiers (Super, Hyper and Meta).

  - Key bindings can be modified in any rc file or interactively using
    the `global-set-key' function.  See notes in `system-gschemrc' for
    examples.

* `gschem' now displays help files and component documentation on all
  platforms, including Windows.  The `locate' tool is no longer used
  to search for component documentation files.

* The `schdiff' tool for generating visual diffs of schematics and
  symbols has been added.  It can be used with most version control
  systems.  See `man schdiff' for more details.

* The `bom' and `bom2' backends to `gnetlist' now give error messages
  if no attribute file can be found.  The `-O attrib_file=FILE' and
  `-O attribs=ATTRIB,ATTRIB...' options have also been added.

* A new `ewnet' backend has been added to `gnetlist'.  This outputs
  netlists for use with the National Instruments ULTIboard layout
  tool.

* The `verilog' backend to `gnetlist' now supports concatenated net
  naming like "{a,b,c[3:0]}".

* `gsch2pcb' no longer supports the `m4_command' parameter in project
  files.

* The `--m4-pcbdir' and `--m4-file' arguments to `gsch2pcb' now work
  again.

* `gsymfix.pl' has been renamed to `gsymfix'.

* Man pages are now available for all programs installed as part of
  gEDA/gaf.

* The undocumented programs `gnet_hier_verilog', `gsymupdate', `gschupdate'
  and `sch2eaglepos.sh' are no longer installed to ${prefix}/bin.

Notable changes in gEDA/gaf 1.7.1
=================================

* gEDA is now compatible with Guile 2.0.


gaf/gaf.c
/*
 * gEDA/gaf command-line utility
 * Copyright (C) 2012 Peter Brett <peter@peter-b.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif
#include <version.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>

/* Gettext translation */
#include "gettext.h"

#include <glib.h>

#include "builtins.h"

#define short_options "+hV"

static struct option long_options[] =
  {
    {"help", 0, NULL, 'h'},
    {"version", 0, NULL, 'V'},
    {"no-rcfiles", 0, NULL, 2},
    {NULL, 0, NULL, 0},
  };

struct internal_command {
  char *name;
  int (*func)(int, char **);
};

static struct internal_command commands[] =
  {
    {"shell", cmd_shell},
    {"config", cmd_config},
    {"export", cmd_export},
    {NULL, NULL},
  };

/* Print help info and exit */
static void
usage (void)
{
  printf (_("Usage: gaf [OPTION...] COMMAND [ARGS ...]\n"
"\n"
"gEDA/gaf command-line utility.\n"
"\n"
"General options:\n"
"  --no-rcfiles   inhibit loading of 'gafrc' files\n"
"  -h, --help     display usage information and exit\n"
"  -V, --version  display version information and exit\n"
"\n"
"Commonly-used commands (type `gaf <cmd> --help' for usage):\n"
"  shell          Scheme REPL for interactive gEDA data processing\n"
"  config         Edit gEDA configuration\n"
"  export         Export gEDA files in various image formats.\n"
"\n"
"Please report bugs to %s.\n"),
PACKAGE_BUGREPORT);
  exit (0);
}

/* Print version info and exit */
static void
version (void)
{
  printf(_("gEDA/gaf %s (g%.7s)\n"
"Copyright (C) 1998-2012 gEDA developers\n"
"This is free software, and you are welcome to redistribute it under\n"
"certain conditions. For details, see the file `COPYING', which is\n"
"included in the gEDA distribution.\n"
"There is NO WARRANTY, to the extent permitted by law.\n"),
         PACKAGE_DOTTED_VERSION, PACKAGE_GIT_COMMIT);
  exit (0);
}

int
main (int argc, char **argv)
{
  int c;
  char *cmd = NULL;
  int cmd_argc = 0;
  char **cmd_argv = NULL;
  int (*cmd_func)(int, char **) = NULL;

  /* Set up gettext */
#if ENABLE_NLS
  setlocale (LC_ALL, "");
  bindtextdomain ("geda-gaf", LOCALEDIR);
  textdomain ("geda-gaf");
  bind_textdomain_codeset ("geda-gschem", "UTF-8");
#endif

  while (-1 != (c = getopt_long (argc, argv, short_options,
                                 long_options, NULL))) {
    switch (c) {

    case 0:
      /* This is a long-form-only flag option, and has already been
       * dealt with by getopt_long(). */
      break;

    case 2: /* --no-rcfiles */
      setenv ("GAF_INHIBIT_RCFILES", "1", 1);
      break;

    case 'V':
      version ();

    case 'h':
      usage ();

    case '?':
      /* getopt_long already printed an error message */
      fprintf (stderr, _("\nRun `gaf --help' for more information.\n"));
      exit (1);
      break;

    default:
      g_assert_not_reached ();
    }
  }

  /* The next argument should be a command */
  if (optind == argc) {
    fprintf (stderr,
             _("ERROR: You must specify a command to run.\n"
               "\n"
               "Run `gaf --help' for more information.\n"));
    exit (1);
  }

  cmd = argv[optind];

  /* Look up the command */
  int i;
  for (i = 0; commands[i].name != NULL; i++) {
    if (strcmp (cmd, commands[i].name) == 0) {
      cmd_func = commands[i].func;
      break;
    }
  }
  if (cmd_func == NULL) {
    fprintf (stderr,
             _("ERROR: Unrecognised command `%s'.\n"
               "\n"
               "Run `gaf --help' for more information.\n"),
             cmd);
    exit (1);
  }

  cmd_argc = argc - optind;
  cmd_argv = argv + optind;
  optind = 1;

  return cmd_func (cmd_argc, cmd_argv);
}
